File: src/domain/strategy/strategyMatrix.ts
import { Hand } from '../models/hand'
import { Card } from '../models/card'
import { BlackjackAction } from '../models/gameState'
import { strategyKeyFromHand, strategyKeyToString } from './strategyKey'
import { buildStrategyTable, StrategyTable } from '../../data/basicStrategyData'

export class StrategyMatrix {
  private static _shared: StrategyMatrix | null = null
  private readonly table: StrategyTable

  private constructor() {
    this.table = buildStrategyTable()
  }

  static get shared(): StrategyMatrix {
    if (!StrategyMatrix._shared) {
      StrategyMatrix._shared = new StrategyMatrix()
    }
    return StrategyMatrix._shared
  }

  getOptimalAction(hand: Hand, dealerUpCard: Card): BlackjackAction {
    if (hand.isBust || hand.isBlackjack) return BlackjackAction.STAND

    const key = strategyKeyFromHand(hand, dealerUpCard)
    const keyStr = strategyKeyToString(key)
    return this.table.get(keyStr) ?? (hand.hardTotal >= 17 ? BlackjackAction.STAND : BlackjackAction.HIT)
  }
}

File: BlackjackApp/Domain/Strategy/StrategyMatrix.swift
import Foundation

final class StrategyMatrix: Sendable {
    static let shared = StrategyMatrix()
    private let table: StrategyTable

    private init() {
        table = BasicStrategyData.table
    }

    /// Returns the optimal Basic Strategy action for the given hand and dealer upcard.
    /// Falls back to .stand for any edge case not in the table (e.g. score > 21).
    func getOptimalAction(for hand: Hand, dealerUpCard: Card) -> BlackjackAction {
        // Busted or Blackjack — no action needed, return stand as sentinel
        guard !hand.isBust && !hand.isBlackjack else { return .stand }

        let key = StrategyKey.from(hand: hand, dealerUpCard: dealerUpCard)
        return table[key] ?? defaultAction(for: hand)
    }

    // MARK: – Private fallback

    private func defaultAction(for hand: Hand) -> BlackjackAction {
        // Sensible default if a key is somehow missing from the table
        hand.bestScore >= 17 ? .stand : .hit
    }
}
